### ISSUE TITLE
Missing Build Instructions

### ISSUE THREAD

**Issue by MartinX3** [MoonchildProductions/FossaMail#31]
How to build fossamail from source?

I want to make a fossa mail PPA for Ubuntu/LinuxMint

**Comment by mattatobin** [comment]
It isn't much different than building Pale Moon just with a few extra steps.. Generally follow older Thunderbird build instructions accounting for Pale Moon backend .. things.

**Comment by MartinX3** [comment]
I tried the build instructions at
[http://developer.palemoon.org/Developer_Guide:Build_Instructions/Pale_Moon/Linux](http://developer.palemoon.org/Developer_Guide:Build_Instructions/Pale_Moon/Linux),
but only until
> Test-running Newly-Built Pale Moon
> 
> Once "./mach build" and "./mach package" have run ruccessfully, you can test-run Pale Moon, assuming that build is not for a different architecture or cpu, which won't run on your machine. 

There is no
> "./mach build" and "./mach package"

But I will search for the Thunderbird v38 build instructions and try these.
I will report back, if there are any problems with them. :)

PS:
Could you send me the right
> .mozconfig
file?
Mine looks like this one, but i don't know, if there is everything correct.
[http://pastebin.com/CDKGdxy6](http://pastebin.com/CDKGdxy6)

**Comment by MartinX3** [comment]
[http://pastebin.com/Wb2cRn7k](http://pastebin.com/Wb2cRn7k)

**Comment by MartinX3** [comment]
[OSError: [Errno 13] Permission denied](http://pastebin.com/s2GPdxFS)

**Comment by wolfbeast** [comment]
Or, instead of creating a post with 25 edits by blindly jabbing at attempts to build without info, you could just be patient and wait for an answer. :)

**Comment by wolfbeast** [comment]
I don't have time right now to fully document it, but the basic steps are:

0. Get some coffee and cookies (optional for some)
1. Check out the FossaMail git
2. Make sure you check out submodules of the repo (Pale Moon will be pulled in as a submodule into `$SRCDIR/mozilla` automatically)
3. Create a proper mozconfig. FTR, Windows release versions use the following right now:
    mk_add_options MOZ_MAKE_FLAGS=-j3
    #palemoon branding
    mk_add_options MOZ_APP_NAME=FossaMail
    mk_add_options MOZ_APP_DISPLAYNAME=FossaMail
    ac_add_options --with-distribution-id=MCP
    #Note: don't use official branding unless you are adhering to the redist license.
    ac_add_options --enable-official-branding
    mk_add_options MOZ_CO_PROJECT=mail
    ac_add_options --enable-application=mail
    ac_add_options --enable-optimize="-O2 -GT -GS- -Qfast_transcendentals"
    ac_add_options --enable-jemalloc
    ac_add_options --disable-debug
    ac_add_options --disable-tests
    ac_add_options --enable-shared-js
    ac_add_options --enable-update-packaging
    ac_add_options --disable-updater
    ac_add_options --disable-devtools
    ac_add_options --disable-webrtc
    ac_add_options --enable-calendar
    WIN32_REDIST_DIR=$VCINSTALLDIR/redist/x86
Linux will probably need a few different things, @trav90 -- can you let him know what is in use for how you build?
4. In your build shell, issue `./mozilla/mach build` <- this is where you didn't manage, because `mach` is part of the platform, not the application code that lives above it, so it can be found in `./mozilla/` instead where the platform code lives.
5. Go have some more coffee.

**Comment by MartinX3** [comment]
Thank you very much for your help :)

I merged your .mozconfig with my one and get the following compiler errors

> 0:04.81 gcc: error: unrecognized command line option '-GT'
 0:04.81 gcc: error: unrecognized command line option '-GS-'
 0:04.81 gcc: error: unrecognized command line option '-Qfast_transcendentals'

**Comment by trav90** [comment]
@MartinX3 the mozconfig you linked to above looks good to me.

And don't forget the coffee!

**Comment by MartinX3** [comment]
@trav90 Thank you for watching over it!
Here is my actual version of it, after the merge with the one from @wolfbeast !
> [http://pastebin.com/ZZ8Xz6JF](http://pastebin.com/ZZ8Xz6JF)

And @trav90 
What is behind:
># ac_add_options --with-external-source-dir=/home/trava90/FossaMail/fmsrc

**Comment by trav90** [comment]
You don't need the update-packaging or gstreamer options.

I believe the --with-external-source-dir is defined and used automagically.

One question I have ESPECIALLY if you plan to redistribute these builds in a PPA - what version of GCC are you using?

**Comment by MartinX3** [comment]
Thank you!

Ahhh, I remember the unstable behaviour of the PaleMoon from Steven Pusser, I made a bug ticket from :D

I see now, the one from your website is built with gcc v4.9.2 and c++ v4.9.2.

My one is build with gcc 5.4.0 and c++ 5.4.0.
Is that okay?

**Comment by mattatobin** [comment]
No. You MUST use 4.9.. 4.7 theoretically works but who knows these days.

**Comment by MartinX3** [comment]
Thank you!

First of all
> ac_add_options --disable-gstreamer

is important on my Linux Mint 18.1 (Based on Ubuntu 16.04), because i can't install the gstreamer development files (version conflicts in the ubuntu 16.04 repo)

Now i made the following steps

> sudo apt-get install g++-4.9 gcc-4.9
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-5 10
sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.9 20
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-5 10
sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-4.9 20
update-alternatives --query g++
update-alternatives --query gcc
./mozilla/mach configure && ./mozilla/mach build

Now, I'm building with gcc-4.9 and g++-4.9 :D

**Comment by wolfbeast** [comment]
GStramer isn't used any longer in our source tree, so the config option doesn't do anything anymore ;)

**Comment by MartinX3** [comment]
Good to know :)

After compiling with gcc 4.9 instead of gcc 5.4 i get the following compiler error :(
> 47:00.25 
47:00.25 /usr/bin/ld: ../../xpcom/components/nsComponentManager.o: relocation R_X86_64_PC32 against Geschütztes Symbol `end_kPStaticModules_NSModule' can not be used when making a shared object
47:00.25 /usr/bin/ld: final link failed: Ungültiger Wert
47:00.25 collect2: Fehler: ld gab 1 als Ende-Status zurück
47:00.25 /media/martin/c33a3882-74f4-4479-8be6-9248cb6934b7/toolchain/Develop/Ubuntu_Packages/FossaMail/git/mozilla/config/rules.mk:818: die Regel für Ziel „libxul.so“ scheiterte
47:00.25 make[4]: *** [libxul.so] Fehler 1
47:00.25 /media/martin/c33a3882-74f4-4479-8be6-9248cb6934b7/toolchain/Develop/Ubuntu_Packages/FossaMail/git/mozilla/config/recurse.mk:74: die Regel für Ziel „toolkit/library/target“ scheiterte
47:00.25 make[3]: *** [toolkit/library/target] Fehler 2
47:00.25 /media/martin/c33a3882-74f4-4479-8be6-9248cb6934b7/toolchain/Develop/Ubuntu_Packages/FossaMail/git/mozilla/config/recurse.mk:36: die Regel für Ziel „compile“ scheiterte
47:00.25 make[2]: *** [compile] Fehler 2
47:00.25 /media/martin/c33a3882-74f4-4479-8be6-9248cb6934b7/toolchain/Develop/Ubuntu_Packages/FossaMail/git/mozilla/config/rules.mk:541: die Regel für Ziel „default“ scheiterte
47:00.25 make[1]: *** [default] Fehler 2
47:00.26 client.mk:404: die Regel für Ziel „build“ scheiterte
47:00.26 make: *** [build] Fehler 2
47:00.26 2 compiler warnings present.

**Comment by wolfbeast** [comment]
Yeah we've had other people running into issues trying to build with a different linker than Gold.
Glad to hear you managed!

Side note:
> ./mozilla/mach clobber && ./mozilla/mach configure && ./mozilla/mach build

You shouldn't have to call `mach configure` manually. `mach build` will invoke it if and when necessary.

**Comment by stevenpusser** [comment]
I have a little bit of code in the debian/rules file for Pale Moon that will automatically use gcc-4.9 and associated programs for the releases that I know use a newer gcc by default.  So far the gcc-4.7 builds for Wheezy also seem to be working:

[code]# Build with gcc-4.9 on Stretch, Xenial, Yakkety, and Zesty.
distrelease := $(shell lsb_release -cs)
ifeq ($(distrelease),$(filter $(distrelease),stretch xenial yakkety zesty))
export CC=gcc-4.9
export CXX=g++-4.9
export CPP=cpp-4.9
export LD=gcc-4.9
endif
[/code]

Then I have the build-depends in debian/control have alternatives for any possible old distros like Wheezy:

[code]# Add for the conditional in rules to force use of gcc-4.9 on newer distreleases
# that default to gcc-5 or 6
        lsb-release,
        gcc-4.9 | gcc-4.8 | gcc-4.7,
        g++-4.9 | g++-4.8 | g++-4.7,
        cpp-4.9 | cpp-4.8 | cpp-4.7,
[/code]

**Comment by stevenpusser** [comment]
I managed to get an experimental build from debianized (adapted from Pale Moon's) source files on my own machine, and plan to set up an repo on the openSUSE build service for Ubuntu 14.04, 16.04, 16.10, 17.04, plus Debian 7 and 8.  If these pass inspection, I'll move it into my Pale Moon OBS project.  I made the source by getting the Fossa Mail source and extracting it, and since I already had a Pale Moon 27.2.1 source tarball on my drive, extracting that, renaming it "mozilla", and putting it into the FM source.  It worked, but maybe there's stuff I can remove from the PM source to reduce the size--let me know if there is!

The mozconfig I used: 
```
export MOZILLA_OFFICIAL=1
mk_add_options MOZ_APP_NAME=FossaMail
mk_add_options MOZ_APP_DISPLAYNAME=FossaMail
ac_add_options --with-distribution-id=MCP
#Note: don't use official branding unless you are adhering to the redist license.
ac_add_options --enable-official-branding
mk_add_options MOZ_CO_PROJECT=mail
ac_add_options --enable-application=mail
ac_add_options --enable-optimize="-O2 -msse2 -mfpmath=sse"
ac_add_options --enable-jemalloc
ac_add_options --disable-debug
ac_add_options --disable-tests
ac_add_options --enable-shared-js
ac_add_options --enable-update-packaging
ac_add_options --disable-updater
ac_add_options --disable-devtools
ac_add_options --disable-webrtc
ac_add_options --enable-calendar

```

**Comment by trav90** [comment]
LGTM

**Comment by MartinX3** [comment]
Thank you, that you will do the work providing a repo for fossamail.
It is better, that PaleMoon and FossaMail will be in the same Repo and with my job and my studies of economics and informatics, there isn't much time for me to invest in this job.

If I can help you, tell me.

**Comment by stevenpusser** [comment]
You could test my builds to see if you can find any problems...

The experimental repo instructions are here:  https://software.opensuse.org//download.html?project=home%3Astevenpusser%3AFossaMail&package=fossamail

The command to add the repo key must also done in a root terminal, this can be attained in Ubuntu with 

"sudo -i".  If the packages pass muster, I can move Fossa Mail into the Pale Moon repository.

**Comment by MartinX3** [comment]
[Erm, the fossamail project will be shut down next month at the same time, a working repo for FM got created?(http://www.fossamail.org/)
That's fate. xD
And bad to hear.

By the way, I can start fossamail and use it with my mail accounts. My profile get's used flawlessy. :)

**Comment by wolfbeast** [comment]
You're free to fork it and continue it under a different name as your own project.

**Comment by MartinX3** [comment]
I would, but it will end in a disaster.
My knowledge is in Java, C#, HTML, CSS, JavaScript.
Not in C++. :)
